language/tools/mirai-dataflow-analysis/README.md

```diff
@@ -0,0 +1,119 @@
+```
+$ cd analyses/
+$ ddlog -i ddlog.dl
+$ cd ddlog_ddlog && cargo build --release
+```
+This will create a Rust crate (`analyses/ddlog_ddlog`) that includes a CLI that the `mirai-dataflow`
+interfaces with to perform its analysis.
+
+## Configuration
+
+`mirai-dataflow` is configured via a combination of CLI arguments and a configuration file.
+
+### CLI
+
+`mirai-dataflow` exposes the following CLI (viewable via the `--help` flag):
+```
+USAGE:
+    mirai-dataflow-analysis [FLAGS] [OPTIONS] <crate-path> <config-path>
+
+FLAGS:
+        --call-graph-only    Only produce a call graph (no analysis)
+    -f, --fresh-build        Perform a rebuild of the crate before analysis
+    -h, --help               Prints help information
+    -V, --version            Prints version information
+
+OPTIONS:
+    -d, --datalog-backend <datalog-backend>            Datalog backend to use (DifferentialDatalog | Souffle)
+    -t, --type-relations-path <type-relations-path>    Path to input type relations
+
+ARGS:
+    <crate-path>     Path to the crate to analyze
+    <config-path>    Path to configuration file
+```
+
+The `crate-path` is the path to the crate you want to analyze. The `config-path` is a path to
+a configuration file (explained below). Finally, the `type-relations-path` is a path to a set of
+input type relations (see the call graph generator's
+[documentation](https://github.com/facebookexperimental/MIRAI/blob/main/documentation/CallGraph.md#type-relations)
+for details).
+
+### Config File
+
+One of the required CLI arguments is a path to a configuration file. This configuration file tells
+`mirai-dataflow` how the generated call graph should be preprocessed before analysis.
+
+The configuration file is a JSON file following this schema:
+```
+{
+    "reductions": List[Reduction],
+    "included_crates": List[String]
+}
+```
+where a Reduction is one of:
+```
+{"Slice": String}
+"Fold"
+"Clean"
+"Deduplicate"
+```
+
+Please see the call graph generator's
+[documentation](https://github.com/facebookexperimental/MIRAI/blob/main/documentation/CallGraph.md)
+for more details.
+
+Below is an example of a valid configuration file:
+```
+{
+    "reductions": [
+        {"Slice": "verify_impl"},
+        "Fold",
+        "Clean"
+    ],
+    "included_crates": ["check_bounds"]
+}
+```
+
+## Usage
+
+Having configured `mirai-dataflow` as explained above, the analysis can then be run:
+```
+$ cargo run -- ../../move-binary-format ./config/call_graph_config.json --datalog-backend=Souffle
+```
+
+`mirai-dataflow` will then execute MIRAI on the crate and then execute the Datalog analysis:
+```
+...
+Running MIRAI...
+Done
+Running analysis...
+Done
+Processing output...
+Done
+```
+
+Finally, a folder called `output` should be created which will contain all of the analysis
+artifacts from this run. Most importantly, it will contain a file `decoded.json` that has
+the decoded Datalog output relation results.
```

Comment:
You probably need a lot more discussion about what ends up in decoded.json, who should look at it and what actions they are supposed to take when the output indicates that there is a problem.

Reply:
Good idea, I've added some text in the readme to address this.